core: set zend_pass_function as default constructor

Girgias opened the pull request:
Currently, it's basically impossible to determine reliably if an internal class is instantiable.

The current way is to overload the `get_constructor` object handler to throw an exception and return `NULL`.
This is problematic for two reasons:
- It's easy to forget to do, compared to simply checking `ce->constructor`
- `get_constructor` can return `NULL` if no explicit constructor was set

This complexity in semantics was one of the motivating factors for introducing the [`object_init_with_constructor()`](https://github.com/php/php-src/pull/14440) API. However, this doesn't cleanly apply to many cases due to the "unorthodox" behaviour of DB drivers prepopulating object properties and being able to modify them in the constructor.

These issues are mainly problematic as it means it's possible to create broken instances of internal objects (e.g. `CurlHandler`).

One other issue with the current design, is that to be able to call the `get_constructor` object handler one must first *create* the object just to immediately discard it, this is wasteful but moreover in those cases the destructor of the object shouldn't run and one must remember to call the `zend_object_store_ctor_failed(obj);` API instead of the usual `OBJ_RELEASE()`.

Therefore, we propose to use the "magic" internal  `zend_pass_function`, which is already used as a dummy constructor call, as the default value for `ce->constructor`. And rather than overriding the `get_constructor` object handler to mark a class as instantiable we add a `\#[NonInstantiableClass("reason")]` attribute that replaces the `ce->constructor` with `NULL`.

This design has a few advantages:
- A class is instantiable if `ce->constructor != NULL`
- It is extensible to userland
- It is possible to remove the `get_constructor` object handler (as it is only overridden to throw a custom exception)

This would also fix certain Reflection bugs such as https://github.com/php/php-src/issues/17796, and should make it finally possible to grab a closure using the FCC syntax of `new MyClass(...)`.

I'm quite happy to not offer this to userland yet while we iron out the kinks (and I don't have the energy to submit an RFC for this at the moment) by making the attribute internal use only.

Remaining planned items:
- [ ] Remove the `get_constructor` object handler.
- [ ] Ensure opcache works as intended
- [ ] Potentially use a class flag instead of needing to always fetch the attribute?

Girgias committed:
core: set zend_pass_function as default constructor

This is to make the semantics of when a class cannot be instiantiated obvious rather than needed to remember to call the get_constructor() handler.

For this we create a new NonInstantiableClass attribute that can be set to mark classes which shouldn't be instantiable with a message and set the constructor pointer to NULL

arnaud-lb commented:
+1 for eventually removing `get_constructor` (or at least making it class-scoped instead of object-scoped by moving it to `zend_class_entry`).

I'm not a fan of using `ce->constructor == NULL` as a proxy for "object is not instantiable" and of setting it to a no-op function by default because this forces the `new` operator for check for it in the happy path.

An alternative would be to add a private constructor that throws on these classes (as suggested in https://github.com/php/php-src/issues/17796#issuecomment-2660382281). This can be done by gen_stubs when a class is annotated with `\#[NonInstantiableClass("reason")]`. The same constructor implementation can be re-used (the reason message can be fetched at runtime from attributes).

This way the `new` operator can simply call the constructor if there is one, and visibility checks will do the rest.

Reflection `::isInstantiable()` can lookup the attribute, but checking that a class has a private constructor may be enough?

Girgias commented:
> +1 for eventually removing `get_constructor` (or at least making it class-scoped instead of object-scoped by moving it to `zend_class_entry`).
> 
> I'm not a fan of using `ce->constructor == NULL` as a proxy for "object is not instantiable" and of setting it to a no-op function by default because this forces the `new` operator for check for it in the happy path.
> 
> An alternative would be to add a private constructor that throws on these classes (as suggested in [#17796 (comment)](https://github.com/php/php-src/issues/17796#issuecomment-2660382281)). This can be done by gen_stubs when a class is annotated with `\#[NonInstantiableClass("reason")]`. The same constructor implementation can be re-used (the reason message can be fetched at runtime from attributes).
> 
> This way the `new` operator can simply call the constructor if there is one, and visibility checks will do the rest.


I'm happy to make the attribute add a private constructor, but my initial understanding as to why nobody wanted to pursue this is because it wouldn't allow a custom message like we currently have. But I guess this is just changing how we handle errors.

And just to clarify, when you say gen_stub can generate the implementation do you want it to generate individually constructor methods, or that it points to a single implementation that is aliased?

> Reflection `::isInstantiable()` can lookup the attribute, but checking that a class has a private constructor may be enough?

This probably could work as most `newInstance` reflection methods check if the constructor is public, but not `newInstanceWithoutConstructor()` so we might need to do a small BC break to normalise its behaviour with the rest of the extension and language.

Might make sense to add a helper function `zend_is_object_instantiable` that does all the checks if it's a concrete class with a public constructor, as the logic is repeated quite often.

Girgias committed:
Use a private constructor instead

Add some APIs so that we can stop relying on the get_constructor object handler

Girgias committed:
Remove get_constructor object handler

arnaud-lb commented on Zend/zend_compile.c:
```diff
@@ -9530,6 +9529,33 @@ static void zend_compile_enum_backing_type(zend_class_entry *ce, zend_ast *enum_
 	zend_type_release(type, 0);
 }
 
+static ZEND_FUNCTION(non_instantiable_constructor)
+{
+}
+
+static zend_arg_info zend_non_instantiable_constructor_arg_info[1] = {0};
+ZEND_API const zend_internal_function zend_non_instantiable_constructor = {
+	ZEND_INTERNAL_FUNCTION, /* type */
+	{0, 0, 0},        /* arg_flags */
+	ZEND_ACC_PRIVATE,     /* fn_flags */
+	NULL,            /* name */
+	NULL,                  /* scope */
+	NULL,               /* prototype */
+	0,                  /* num_args */
+	0,            /* required_num_args */
+	zend_non_instantiable_constructor_arg_info + 1, /* arg_info */
+	NULL,               /* attributes */
+	NULL,        /* run_time_cache */
+	NULL,            /* doc_comment */
+	0,                        /* T */
+	0,                   /* fn_flags2 */
+	NULL,                /* prop_info */
+	ZEND_FN(non_instantiable_constructor), /* handler */
+	NULL,                 /* module */
+	NULL,     /* frameless_function_infos */
+	{NULL,NULL,NULL,NULL}   /* reserved */
+};
```
Possibly this can be generated with a `@generate-function-entries` stub (like `ext/zend_test/tmp_methods.stub.php`).

Girgias replied:
The special pass function is also not generated, so not sure if this one should.

arnaud-lb replied:
It could reduce maintenance later if generating works. With `@generate-function-entries` it generates only the `zend_function_entry`, without any registering code.

Girgias replied:
But don't we need the registering code anyway? As the pass function also does this.

arnaud-lb replied:
Oh yes you are right. I forgot that stubs generate zend_function_entries, not zend_internal_functions.

arnaud-lb commented on Zend/zend_compile.c:
```diff
@@ -9651,6 +9677,14 @@ static void zend_compile_class_decl(znode *result, const zend_ast *ast, bool top
 	}
 
+	/* Add zend_non_instantiable_constructor constructor if class cannot be manually instantiated */
+	if (ce->constructor == NULL && (ce->ce_flags & (ZEND_ACC_ENUM|ZEND_ACC_INTERFACE|ZEND_ACC_TRAIT)) == 0) {
```
Should you error if there is already a constructor? Otherwise we can do this:

```
#[NonInstantiableClass("test")]
class C {
    public function __construct() {}
}

new C(); // works
```

Girgias replied:
I might remove the possibility for userland classes to add this attribute for now as I don't want to go through an RFC.

arnaud-lb replied:
Makes sense. In that case, will you need to hide the `NonInstantiableClass` symbol entirely? (or use an `@` annotation in stubs instead?)

Girgias replied:
Should we be hiding the symbol? I think it could still be useful to expose it for people?

Maybe @TimWolla or @DanielEScherzer have some opinions (or willingness to write the RFC for it)

arnaud-lb commented on NEWS:
```diff
@@ -126,9 +126,8 @@ PHP                                                                        NEWS
-  . pg_fetch_object() now surfaces non-instantiable class errors
-    before fetching, resolves the constructor via the get_constructor
-    handler, and reports the empty-constructor ValueError on the
+  . pg_fetch_object() now surfaces non-instantiable class errorsv before
```
``` suggestion
  . pg_fetch_object() now surfaces non-instantiable class errors before
```

arnaud-lb commented on Zend/zend_objects_API.c:
```diff
@@ -211,3 +211,70 @@ ZEND_API ZEND_COLD zend_property_info *zend_get_property_info_for_slot_slow(zend
+	}
+}
+
+ZEND_API bool zend_check_class_is_instantiable_or_throw(const zend_class_entry *ce, const zend_class_entry *scope) {
```
From the context of `ZEND_NEW` there is some redundancy as some of these checks are also performed by `object_init_ex`. We could rename this method to "check_constructor_visibility_or_throw", remove the `ZEND_ACC_UNINSTANTIABLE` check, and rely on the `non_instantiable_constructor` to throw when it's called.

Girgias replied:
The motivation here is to have a way to determine if a class is instantiable _prior_ to calling the `object_init_ex()` function. I'll introduce a `object_init_instantiable_class()` function that doesn't duplicate the `ZEND_ACC_UNINSTANTIABLE` check.

arnaud-lb replied:
In the context of `ZEND_NEW` it's not necessary to check prior to calling `object_init_ex()`, as this function can fail already. We could also avoid the `zend_is_non_instantiable_constructor()` check if that constructor throws when it's called. This would be faster in the usual case and maybe slower in the case that usually doesn't happen.

Girgias replied:
I'm not fully certain understand how this could be achieved? Do you mean just call the constructor without checking and the fact it is private would make it throw anyway?

arnaud-lb replied:
What I had in mind is that the body of the non_instantiable_constructor could call `zend_cannot_instantiate_class()`.

Girgias replied:
Ohhhhhh right yeah that would actually be smart. And should be quite easy to do :)

arnaud-lb commented on UPGRADING:
```diff
@@ -69,13 +77,19 @@ PHP 8.6 UPGRADE NOTES
 
+- Reflection:
+  . ReflectionClass::newInstanceWithoutConstructor() no longer accepts
+    classes with non-public constructors.
```
What is the reason for this? This provides no extra safety to internal classes as sub-classing does not inherit NonInstantiableClass.

Girgias replied:
Actually no longer the case in the current implementation

Girgias committed:
Address some review concerns

Still missing preventing userland classes from applying the attribute

Girgias committed:
Validate attribute usage

Girgias committed:
Prevent attribute to be used on userland classes

Girgias committed:
Throw the exception inside the non instantiable constructor
